src/Razor/test/Microsoft.VisualStudio.Razor.Integration.Test/GoToDefinitionTests.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Xunit;
+
+namespace Microsoft.VisualStudio.Razor.Integration.Test
+{
+    public class GoToDefinitionTests : AbstractRazorEditorTest
+    {
+        [IdeFact]
+        public async Task GoToDefinition_MethodInSameFile()
+        {
+            // Open the file
+            await TestServices.SolutionExplorer.OpenFileAsync(BlazorProjectName, CounterRazorFile, HangMitigatingCancellationToken);
+
+            // Wait for classifications to indicate Razor LSP is up and running
+            await TestServices.Editor.WaitForClassificationAsync(HangMitigatingCancellationToken);
+            await TestServices.Editor.PlaceCaretAsync("IncrementCount", -1, HangMitigatingCancellationToken);
+
+            // Act
+            await TestServices.Editor.InvokeGoToDefinitionAsync(HangMitigatingCancellationToken);
+
+            await TestServices.Editor.WaitForCaretMoveAsync(HangMitigatingCancellationToken);
+
+            // Assert
+            await TestServices.Editor.VerifyCurrentLineTextAsync("private void IncrementCount()", HangMitigatingCancellationToken);
+        }
+
+        [IdeFact]
+        public async Task GoToDefinition_CSharpClass()
+        {
+            // Open the file
+            await TestServices.SolutionExplorer.OpenFileAsync(BlazorProjectName, IndexRazorFile, HangMitigatingCancellationToken);
+
+            // Change text to refer back to Program class
+            await TestServices.Editor.SetTextAsync(@"<SurveyPrompt Title=""@nameof(Program)", HangMitigatingCancellationToken);
+            await TestServices.Editor.PlaceCaretAsync("Program", -1, HangMitigatingCancellationToken);
+
+            // Wait for classifications to indicate Razor LSP is up and running
+            await TestServices.Editor.WaitForClassificationAsync(HangMitigatingCancellationToken);
```

Comment:
Hmm, if we use this as an indicator for "lsp is ready" I wonder if we should add a helper method to the base class which is simply `WaitForServerReady` which does this under the covers. That way if the requirement of "is ready" changes we can update it everywhere in one go.

Reply:
Yeah, I had the same thought but wasn't quite sure. If you agree, then thats good enough for me, I'll do it!

Reply:
Actually.. there was a reason for this.. this only works if there is at least one component in the input file, since it waits specifically for that classification to be seen.. so for the wrong input file, this would just wait forever then fail the test.

Reply:
> Actually.. there was a reason for this.. this only works if there is at least one component in the input file, since it waits specifically for that classification to be seen.. so for the wrong input file, this would just wait forever then fail the test.

Oh geez, that's an interesting implication. Definitely feels like a potential pit folks could fall into unintentionally after copy & pasting tests. Sounds like we may want to revisit how this checks "is ready" in the future which I think warrants the refactoring into its own method too.

Reply:
> WaitForServerReady could open up a known-Component-containing file itself, wait for classification, then close that file to avoid the wait forever scenario.
> Actually, it might behoove us to make WaitForServerReady part of the "Before test" stop so that we don't have to include WaitForServerReady in every test.

Love this idea <3

Will look into it